src/gps/gps/rtcm_pub_node.py

```diff
@@ -0,0 +1,239 @@
+        worker (serial.Serial): Serial connection instance.
+        ubr (UBXReader): UBXReader instance for parsing RTCM/UBX data.
+    """
+
+    def __init__(self, port="/dev/ttyACM0", baud=38400):
+        """
+        Initializes the serial connection and UBXReader.
+
+        Args:
+            port (str): Serial port to connect to.
+            baud (int): Baud rate for the serial connection.
+
+        Raises:
+            RuntimeError: If the serial port cannot be opened.
+        """
+        self.lock = threading.Lock()
+        with self.lock:
+            try:
+                self.worker = serial.Serial(port, baud, timeout=1)
+            except serial.SerialException as e:
+                raise RuntimeError(f"Failed to open serial port {port}: {e}") from e
+            self.ubr = UBXReader(
+                self.worker,
+                protfilter=RTCM3_PROTOCOL,
+                validate=VALCKSUM,
+            )
+
+    def read(self) -> tuple:
+        """
+        Reads data from the serial port using UBXReader.
+
+        Returns:
+            tuple: Raw data and parsed data from UBXReader.
+
+        Raises:
+            RuntimeError: If reading from the serial port fails.
+        """
+        with self.lock:
+            try:
+                return self.ubr.read()
+            except Exception as e:
+                raise RuntimeError(f"Error reading from serial port: {e}") from e
+
+    def write(self, data: bytes):
+        """
+        Writes data to the serial port.
+
+        Args:
+            data (bytes): Data to write to the serial port.
+
+        Raises:
+            RuntimeError: If writing to the serial port fails.
+        """
+        with self.lock:
+            try:
+                self.worker.write(data)
+            except Exception as e:
+                raise RuntimeError(f"Error writing to serial port: {e}") from e
+
+    def data_available(self) -> bool:
+        """
+        Checks if there is data available to read from the serial port.
+
+        Returns:
+            bool: True if data is available, False otherwise.
+        """
+        with self.lock:
+            return self.worker.in_waiting > 0
+
+
+class RtcmNode(Node):
+    """
+    ROS 2 node for managing RTCM data via serial communication.
+
+    Attributes:
+        rtcm_pub (Publisher): Publishes RTCM messages to the /rtcm topic.
+        serial_conn (IoManager): Manages serial I/O operations.
+        layers (int): Configuration layers for the UBXMessage.
+        timer (Timer): Timer for periodically reading and publishing RTCM data.
+    """
+
+    def __init__(self):
+        """
+        Initializes the RTCM node, loads parameters, sets up serial communication,
+        configures RTCM output, and starts a periodic timer callback.
+        """
+        super().__init__("rtcm_node")
+        self.load_params()
+        self.rtcm_pub = self.create_publisher(Rtcm, "/rtcm", 1)
```

Comment:
Magic number usage for what is the timeout i think? , maybe we can create a named constant?

Reply:
It's actually queue depth. I can make it configurable through parameters although we will probably leave it at 1.